keps/sig-storage/3756-volume-reconstruction/README.md

```diff
@@ -0,0 +1,964 @@
+  This DSWP is marked as `hasAddedPods=true` ("fully populated") only after
+  it has read all Pods from files (static pods) **and** the API server (i.e.
+  [`sourcesReady.AllReady` returns `true` here](https://github.com/kubernetes/kubernetes/blob/cca3d557e6ff7f265eca8517d7c4fa719077c8d1/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L150-L159)).
+
+Both ASW and DSW caches exist only in memory and are lost when kubelet process
+dies. It's relatively easy to populate DSW - just list all Pods from the API
+server and static pods and collect their volumes. Populating ASW is complicated
+and actually source of several problems that we want to change in this KEP.
+
+*Volume reconstruction* is a process where kubelet tries to create a single
+valid `PersistentVolumeSpec` or `VolumeSpec` for a volume from the OS.
+Typically from mount table by looking at what's mounted at 
+`/var/lib/kubelet/pods/*/volumes/XYZ`. This process is imperfect,
+it populates only `(Persistent)VolumeSpec` fields that are necessary to unmount
+the volume (i.e. to call `volumePlugin.TearDown` + `UnmountDevice` calls).
+
+Today, kubelet populates VolumeManager's DSW first, from static Pods and pods
+received from the API server. ASW is populated from the OS
+after DSW is fully populated (`hasAddedPods==true`) and **only volumes missing
+in DSW are added there**. In other words, kubelet reconstructs only the volumes
+for Pods that were running, but were deleted from API server before kubelet
+started. (If the pod is still in the API server, Running, its volumes would be
+in DSW).
+
+We assumed that this was enough, because if a volume is in DSW, the
+VolumeManager will try to mount the volume, and it will eventually reach ASW. 
+
+We needed to add
+[a complex workaround](https://github.com/kubernetes/kubernetes/pull/110670)
+to actually unmount a volume if it's initially in DSW, but user deletes all
+Pods that need it before the volume reaches ASW. 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose to reverse the kubelet startup process.
+
+1. Quickly reconstruct ASW from the OS and add **all** found volumes to ASW
+   when kubelet starts as *uncertain*. "Quickly" means the process should look
+   only at the OS and files/directories in `/var/lib/kubelet/pods` and it should
+   not require the API server or any network calls. Esp. the API server may
+   not be available at this stage of kubelet startup.
+2. In parallel to 1., start DSWP and populate DSW from the API server and
+   static pods.
+3. When connection to the API server becomes available, complete reconstructed
```

Comment:
Will the data from OS in 1 have conflicts with data from 3? Which one should we treat as source of truth?

Reply:
Added a note about this to the KEP. `devicePath` reconstructed from the OS is replaced by `devicePath` from node.status.

(We can argue if it's correct or wrong... )